File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:

    ListNode* reverse(ListNode* temp, ListNode* head){
        if(!temp->next){
            head = temp;
            return head;
        };
        head = reverse(temp->next, head);
        temp->next->next = temp;
        temp->next = NULL;
        return head;
    }

    ListNode* reverseList(ListNode* head) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        if(!head) return head;
        return reverse(head, head);
    }
};

File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # if head == None:
        #     return head

        # prev_ptr = None
        # while(head):
        #     temp = head.next
        #     head.next = prev_ptr
        #     prev_ptr = head
        #     head = temp

        # return prev_ptr

        if head == None:
            return head

        if head.next == None:
            return head
        
        new_head = self.reverseList(head.next)
        head.next.next = head
        head.next = None

        return new_head







